### ISSUE TITLE
Is loadCSS still a best practice?

### ISSUE THREAD

**Issue by contolini** [ascott1/ethical-web-dev#39]
Nowadays I think there's actually a much easier way to prevent CSS from blocking rendering, using [inline media queries](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/render-blocking-css?hl=en):

```
<link href="style.css" rel="stylesheet">
<link href="print.css" rel="stylesheet" media="print">
<link href="other.css" rel="stylesheet" media="(min-width: 40em)">
```

> For example, our first stylesheet declaration does not provide any media type or query, hence it will apply in all cases - that is to say, it is always render blocking. On the other hand, the second stylesheet will only apply when the content is being printed - perhaps you want to rearrange the layout, change the fonts, etc - and hence this stylesheet does not need to block the rendering of the page when it is first loaded. Finally, the last stylesheet declaration provides a “media query” which is executed by the browser: if the conditions match, the browser will block rendering until the stylesheet is downloaded and processed.

https://github.com/ascott1/ethical-web-dev/blob/master/web-apps-that-work-everywhere/04-web-performance.md#optimize-the-rendering-path

**Comment by ascott1** [comment]
This is a tricky one. I actually really like the pattern you're describing particularly since there is no JS dependency or shimming involved.

Where that technique loses out is on the "[app shell](https://developers.google.com/web/updates/2015/11/app-shell)" design pattern, which is pretty compelling, since any resource at a used media query will still be render blocking. Google seems to still recommend the delayed CSS loading: https://developers.google.com/speed/docs/insights/OptimizeCSSDelivery#example

Here's a demo of app-shell using a custom loader, similar to loadCSS https://github.com/GoogleChrome/application-shell/blob/master/server/views/partials/async-css.handlebars

**Comment by ascott1** [comment]
All of that said, I think this section might deserve a retooling.

**Comment by ascott1** [comment]
_Here's a draft of how I changed this section. Let me know if you have any feedback @contolini:_

Similarly, for an additional performance boost critical CSS can be loaded inline while load secondary styles are loaded in a non-blocking manner. Critical CSS can be considered anything that is absolutely visually necessary for the initial page load such as baseline typographic styles, color, header styles, and basic layout. I often think of these things as the “shell” of the site or application. If you are working on an existing site, the tool [critical](https://github.com/addyosmani/critical) automatically extracts critical css from our styles. 

Once we have separated our critical and non-critical CSS we can either use media queries and media types to load a small file of necessary CSS or asynchronously load our other styles. Let’s look at both techniques.

In the Responsive Design chapter we explored the use of media queries to selectively style our site at different breakpoints. We are also able to use media queries and media types within the   `<link>` tags that point to our CSS files. When we do this, only the CSS files that apply to browser conditions will be render blocking. Other CSS files will download, but will do so asynchronously. This means that we can significantly decrease the file size of the CSS that the browser needs to download on the first render. To do this we would break our CSS into smaller files and add a `media` attribute to the `<link>` tags:

```
<head>
  <!— inline critical CSS —>
  <style>
    /* inlined critical styles */
  </style>

  /* referenced CSS files */
  /* applies only when a site is being printed */
  <link href="print.css" rel="stylesheet" media="print">
  /* styles applied when viewport is 600px or less */
  <link href="small.css" rel="stylesheet" media="(max-width: 600px)">
  /* styles applied when viewport is 601px or more */
  <link href="large.css" rel="stylesheet" media="(min-width: 601px)">
</head>
```

If we want to explore the asynchronous loading of CSS, Filament Group’s loadCSS provides a means for loading styles based on the preload pattern[^x]. Building upon our JavaScript loading example, we could handle styles in this way:

```
<html>
  <head>
    <!— inline critical CSS —>
    <style>
      /* inlined critical styles */
    </style>

    <!— preload pattern for loading external CSS —>
    <link rel="preload" href="path/to/mystylesheet.css" as="style" onload="this.rel='stylesheet'">
    <!— no js fallback —>
    <noscript><link rel="stylesheet" href="path/to/mystylesheet.css"></noscript>

    <!— inline critical JS —>
    <script type="text/javascript">
      /* inlined critical javascript */
      /* includes filament group's loadCSS */
      /* https://github.com/filamentgroup/loadCSS */
    </script>
  </head>
  <body>
    <h1>Page Content</h1>

    <!— place additional JS file just above /body tag —>
    <script type="text/javascript" src="external.js"></script>
  </body>
</html>
```

As an alternative to loadCSS, [Google’s app-shell demo](https://developers.google.com/speed/docs/insights/OptimizeCSSDelivery#example) includes a custom asynchronous style loader with a very small footprint.

<aside>
**Note:**

Both loadCSS and Google’s app-shell CSS loader require JavaScript to load the secondary CSS file. As long as our inline styles include our site's core styles, this will provide a usable, progressively enhanced experience for our users should JavaScript fail on a page. If you choose to use one of these techniques, I recommend testing your site with only the inlined CSS styles to ensure it remains usable.
</aside>

By considering how our styles and scrips effect the rendering our page, we can increase the time to first render for our user’s.

[^x]: To learn more about preload, I recommend reading the post Preload: What Is It Good For? by Yoav Weiss https://www.smashingmagazine.com/2016/02/preload-what-is-it-good-for/
